### ISSUE TITLE
RFC: New nteract Toolbar Design

### ISSUE THREAD

**Issue by captainsafia** [nteract/archived-desktop-app#4809]
Hello nteract community members!

Over the past few months, we've been working on redesigning the nteract desktop and web app UX based on feedback that users have provided. We'll start posting these design's here to gather feedback from users before we implement them. We'll create a separate issue for each design element and gather feedback for them one-by-one. First up: a new global toolbar.

<hr/>

![Untitled](https://user-images.githubusercontent.com/1857993/71832341-951db580-305f-11ea-88a7-06bdca66683a.png)

The Toolbar provides a faster and more visual way to perform common actions in nteract without relying on shortcuts or digging through menu bar items.

# Whom is this for?

👶 **First-time** users will appreciate a very obvious entry point for most used actions within nteract. It also helps to better understand how nteract works by highlighting typical/frequent actions.

👦 **Occasional users** appreciate the toolbar as it provides often used actions in a highly accessible way. The assumption here is that they might not use nteract often enough to having memorise the shortcuts.

👵 **Veterans** who use nteract on a regular basis might not appreciate the toolbar and rather prefer having more space for actual document contents. Therefore, they want to hide it.

# Functionality

![Kapture_2019-12-24_at_9 18 09](https://user-images.githubusercontent.com/1857993/71832305-87683000-305f-11ea-9cb7-dfb15fdea098.gif)

**Add Cell** adds a code cell after the currently selected cell. Also addresses #2953 and #4458 as this makes it much easier to add a (code) cell

**Restart & Run, Interrupt** and **Clear Outputs** are functionalities which can be found in different places in the app menu. As they are important (and document-wide) functionalities they can now be found in the toolbar

![Kapture_2019-12-24_at_10 33 29](https://user-images.githubusercontent.com/1857993/71832282-7c150480-305f-11ea-94cb-6cbf263cc6b4.gif)

**Hide/Show** allows users to control the visibility of the toolbar

[→ Figma Click Prototype](https://www.figma.com/proto/XIQtlBkUwAF1qso9cOStAm/nteract-Redesign?node-id=241%3A287&viewport=751%2C278%2C0.1862417310476303&scaling=min-zoom)

*Functionalities: Scrollable, Hover/Click "Add", Show/Hide Toolbar*

# Questions

- What is your overall impression?
- Are you happy with how showing/hiding the toolbar works?
- Do you agree with the current selection of toolbar items? (Add Cell, Restart & Run, Interrupt, Clear Outputs)

**Comment by captainsafia** [comment]
Bringing in feedback from Twitter.

https://twitter.com/fizzdoug/status/1214340794737463296

**Comment by gnestor** [comment]
Thanks for the thorough overview @captainsafia! 

My feedback:

* The toolbar could be much more compact. I think that labels are nice for first-time users but they take up a lot of space. Tooltips could suffice. Tooltips with keyboard shortcuts are very helpful.
* The icon for "interrupt" should probably be a stop icon because when the user clicks "Restart & Run," it's not resuming from where it was _paused_, it's running from the beginning. 
* It would be nice-to-have MacOS touch bar support so that the toolbar can be displayed in the touch bar vs. in the window: https://electronjs.org/docs/api/touch-bar
* It would also be nice-to-have toolbar customization so that any command can be added/removed from the toolbar.

**Comment by captainsafia** [comment]
@gnestor Thanks for the feedback! Your point about adding macOS touch bar support to the desktop app is a good one.

When you say toolbar customizations are you thinking of a drag-and-drop experience similar to the kind that some browsers (I'm thinking of Safari) allow for editing the toolbar?

@fabric-8 Consider this two points of feedback for the toolbar being more compact.

**Comment by gnestor** [comment]
@captainsafia Drag-and-drop would be the most user-friendly but obviously much more difficult to implement. JupyterLab has implemented a JSON editor for editing settings which is much easier to implement. Regardless of the specific UI, it should certainly be customizable from within the app vs. editing config files from the terminal/code editor.

**Comment by Jaykul** [comment]
I came to say most of the things @gnestor said in his first post 😉 but I wanted to add a couple of points:
- Could we have multiple views for the toolbar? I.e.: wide, thin, none
- I think the current widely-spaced view with the text underneath (maybe even with a second line showing the keyboard shortcut) would be great on first use ... **but** it would also be really nice for people actually _using_ touch screens instead of a mouse/touchpad (similar to how firefox's UI adds spacing when you interact with it via touch?)

**Comment by Zsailer** [comment]
This looks great, @captainsafia! 😃 Also, kudos for making this easy to review (I love the GIFs, images, and Figma prototype combination—I know that takes a lot of time to present effectively).

My first two thoughts:

1. Are y'all also exploring adding an option to change kernels to this toolbar? I know this probably adds a bit of cognitive load for users (especially if they're new users). The main use-case I have in mind is allowing users to easily change languages through this UI. Exposing all kernelspecs is likely too much for nteract's minimal design, but maybe adding the option for languages that launch from the base spec for each language might be enough/not-too-much.

2. Piggy-backing off @Jaykul's idea:

    > Could we have multiple views for the toolbar?

    I like the idea of giving the user an option to switch to a vertical sidebar. Often times, I have more horizontal screen real estate than vertical, and most of my horizontal space remains un-used in single document views. A vertical sidebar allows me to leverage this UI without sacrificing vertical real estate. 

    The Cal Poly Interns explored a vertical sidebar in their design of a [JupyterLab "focus mode"](https://discourse.jupyter.org/t/focus-mode-intern-project/1552/11) over the summer. I thought I'd share their design prototype here in case you found it useful. 

    ![](https://aws1.discourse-cdn.com/standard11/uploads/jupyter/original/2X/3/31725669249cf9208d6ce01410a2d91eaba0449d.png)


Great work here! Really excited to see the final product! 🎉

**Comment by captainsafia** [comment]
Thanks for the feedback and keep it coming folks!

Our designer, @fabric-8, will be looking at this feedback and incorporating it into the design later this week.

**Comment by fabric-8** [comment]
**Thanks a lot for the feedback everybody! 🎉**

Here's an update considering the following aspects:
> The toolbar could be much more compact. I think that labels are nice for first-time users but they take up a lot of space. Tooltips could suffice. Tooltips with keyboard shortcuts are very helpful.

> The icon for "interrupt" should probably be a stop icon because when the user clicks "Restart & Run," it's not resuming from where it was paused, it's running from the beginning.

![Kapture 2020-01-16 at 10 50 13](https://user-images.githubusercontent.com/4112698/72514164-7634d580-384e-11ea-9a4a-3ea286fd1a8f.gif)

 [🕹 Figma Hover Prototype](https://www.figma.com/proto/XIQtlBkUwAF1qso9cOStAm/nteract-Redesign?node-id=316%3A8886&viewport=36%2C-611%2C0.46697255969047546&scaling=min-zoom)

The timing of the tooltips is not ideal here, but hopefully you get the idea (I would use delays similar to slack to make the appearing / disappearing smoother)

I also do think that customizing the toolbar is a good idea @gnestor @Jaykul , however I'd propose to go for a "basic" implementation first and then listen to feedback and learn more about how people use and perceive the toolbar before moving on – Does that makes sense?

**Comment by jasongrout** [comment]
I love seeing the work that is going into the design of this toolbar. I also like how you are bringing  good document-wide practices to the forefront, like restart and run all and clear outputs.

Do you think there might be a way to indicate on the play button that it will do a restart and play from the beginning, i.e., a global document play, rather than just running the current cell? Here are some thoughts:

* a vertical line before the play like <img width="26" alt="Screen Shot 2020-01-16 at 8 16 37 AM" src="https://user-images.githubusercontent.com/192614/72542257-b0937800-3838-11ea-9ad1-1ad592656666.png"> to indicate that it will stop first and play from the beginning. The line kind of reminds the user of a stop, and the line also indicates we are playing from the start.
* A circular border around the play symbol, like [this](https://fontawesome.com/icons/play-circle?style=solid) or [this](https://fontawesome.com/icons/play-circle?style=solid), or perhaps a rectangular border. The border would be reminiscent of the document, to remind the user that this is a document-level action.
* Or perhaps even just a slightly fancier play button, like [this](https://fontawesome.com/icons/play?style=duotone), to indicate this play is a richer one that just running a single cell.

Both the notebook and JLab use the standard play button to run the current cell. I would love to see a similar toolbar button in those interfaces to help make a suggested good practice common, and it would be great if all three interfaces could standardize on a common icon to represent this.

 CC also @tgeorgeux

**Comment by captainsafia** [comment]
> A circular border around the play symbol, like this or this, or perhaps a rectangular border. The border would be reminiscent of the document, to remind the user that this is a document-level action.

I like this idea the most for a Restart + Run All icon. Something similar to an icon like this.

![image](https://user-images.githubusercontent.com/1857993/72556398-5bb12b00-3853-11ea-8a3c-a1cd97131fc8.png)

With regard to icons, I think this is something we can incrementally iterate on as we roll out the new UI. We can start off with a sensible default and evolve it as we gather more feedback from users during their day-to-day interactions with the nteract desktop/web apps.

**Comment by jasongrout** [comment]
Wow, I like that idea of a restart circle border!

@ivanov also loves the restart around a play, but points out that the arrow is facing the wrong way (check out the reload button on your browser). He says in Western culture, time flows clockwise around a clock (and indeed in sundials, etc. :).

**Comment by jasongrout** [comment]
Though @ivanov also points that maybe "rewind time" is actually the effect you are going for, so perhaps a backwards "refresh" icon is warranted?

**Comment by tgeorgeux** [comment]
Looks a LOT better with the labels gone and the hovers. Especially the shortcuts, really clear, very well executed. 

I agree with making the bar a little more compact, in the latest iteration posted, the buttons being aligned with the edge of the material below makes them look like they're not aligned correctly, I would treat the outside edge of a notebook cell as the edge content on the page, and the left-most button some padding. This ties the buttons more to the cells below.

One thing sticking out to me right now is we have three types of buttons showing on the screen right now and they're all different sizes and they don't have uniform padding/margins. I would size all the buttons on the toolbar to match the existing cell-specific controls. I would also adjust build in a consistent set (or two) of padding around buttons, and use those throughout.

I am not sure that filled buttons are necessary. How often do you expect the toolbar to be used? Right now it's taking a lot of attention by having so much saturation compared to the rest of the UI. I would play around with flat buttons.

TL:DR
- The toolbar could be even more compact.
    - Treat the width of the cells as the left of the UI
- Make uniform button sizes (or at least match the size/padding of the cell UI controls and the open/collapse icon).
- Explore flat buttons.


Here's a basis for the a 'restart and run' Icon.
![--jp-icon-restart-and-run](https://user-images.githubusercontent.com/22109376/72565065-f0248900-3865-11ea-9259-46d67bb3c575.png)

And one with rounded edges to match the others. 
![--jp-icon-restart-and-run-round](https://user-images.githubusercontent.com/22109376/72567895-b6568100-386b-11ea-9dec-f1bbf3b56f16.png)

**Comment by pdimens** [comment]
> This looks great, @captainsafia!  Also, kudos for making this easy to review (I love the GIFs, images, and Figma prototype combination—I know that takes a lot of time to present effectively).
> 
> My first two thoughts:
> 
> 1. Are y'all also exploring adding an option to change kernels to this toolbar? I know this probably adds a bit of cognitive load for users (especially if they're new users). The main use-case I have in mind is allowing users to easily change languages through this UI. Exposing all kernelspecs is likely too much for nteract's minimal design, but maybe adding the option for languages that launch from the base spec for each language might be enough/not-too-much.
> 2. Piggy-backing off @Jaykul's idea:
>    > Could we have multiple views for the toolbar?
>    
>    
>    I like the idea of giving the user an option to switch to a vertical sidebar. Often times, I have more horizontal screen real estate than vertical, and most of my horizontal space remains un-used in single document views. A vertical sidebar allows me to leverage this UI without sacrificing vertical real estate.
>    The Cal Poly Interns explored a vertical sidebar in their design of a [JupyterLab "focus mode"](https://discourse.jupyter.org/t/focus-mode-intern-project/1552/11) over the summer. I thought I'd share their design prototype here in case you found it useful.
>    ![](https://camo.githubusercontent.com/4213e14119e77bfa606eda853997c7ff12b193bb/68747470733a2f2f617773312e646973636f757273652d63646e2e636f6d2f7374616e6461726431312f75706c6f6164732f6a7570797465722f6f726967696e616c2f32582f332f333137323536363932343963663932303864366365303134313061326439316561626130343439642e706e67)
> 
> Great work here! Really excited to see the final product!

I really like this compact implementation. If people wanted to preserve their horizontal space, then this little toolbar can possibly float just above the active cell, maybe allowing some overlap with the previous cell since the focus is not on that cell anymore.

**Comment by captainsafia** [comment]
We shipped this. Closing.
